audinterface/core/feature.py

```diff
@@ -301,25 +301,58 @@ def process_index(
             index: pd.Index,
             *,
             root: str = None,
+            cache_root: str = None,
     ) -> pd.DataFrame:
         r"""Extract features from an index conform to audformat_.
 
         Args:
             index: index with segment information
             root: root folder to expand relative file paths
+            cache_root: cache result under this folder.
+                Filename will be created with
+                :func:`audformat.utils.hash`
 
         Raises:
             RuntimeError: if sampling rates do not match
             RuntimeError: if channel selection is invalid
             RuntimeError: if multiple frames are returned,
                 but ``win_dur`` is not set
             ValueError: if index is not conform to audformat_
+            ValueError: if cached file has different column names
 
         .. _audformat: https://audeering.github.io/audformat/data-format.html
 
         """
-        series = self.process.process_index(index, root=root)
-        return self._series_to_frame(series)
+        cache_path = None
+
+        if cache_root is not None:
+            cache_root = audeer.mkdir(cache_root)
+            hash = audformat.utils.hash(index)
+            cache_path = os.path.join(cache_root, f'{hash}.pkl')
+
+        if cache_path and os.path.exists(cache_path):
+
+            df = pd.read_pickle(cache_path)
+            if list(df.columns) != self.column_names:
```

Comment:
There is of course still the risk that you run a different model with the same output names. In that case a user would not notice. To make it safer we could process the first segment of the index and compare it with the cached values.

Reply:
I actually decided to implement it that way. Now not only the columns have to match, but also the values of the first segment. It's a bit safer that way.

Reply:
The downside is that you need to be able to run the model, even when loading from cache. E.g. if the model can only be run on GPU and no GPU is available, it would be more convenient if we don't have to process the first segment.

I think it would be safer if we could achieve something like `<cache-root>/<model>/<hash>.pkl`. The question is then how to get `<model>`. One way is that a user might be providing it by an attribute or an argument when instantiating the `audinterafce` object. We might fall back to a standard name if we cannot get it, but it would still be nice, if every model that is loaded with `audmodel` or `audonnx` could automatically result in some model name.

Reply:
> I think it would be safer if we could achieve something like `<cache-root>/<model>/<hash>.pkl`. The question is then how to get `<model>`. One way is that a user might be providing it by an attribute or an argument when instantiating the `audinterafce` object. We might fall back to a standard name if we cannot get it, but it would still be nice, if every model that is loaded with `audmodel` or `audonnx` could automatically result in some model name.

But `audinterface` is meant to work with any processing function not just `audonnx` models.

Reply:
I know, in that case we can still use a fallback name for `<model>`.

Reply:
> There is of course still the risk that you have a different process function but use the same feature names

This will happen all the time. If you would like to compare different gender models, you would always call the feature gender. Or even don't use `audinterface.Feature` but `audinterface.Process` to get the values back as a Series.

Reply:
Mhh, not convinced :) Then I would rather skip the additional test.

Reply:
I would also skip the additional test, but I would like to have the possibility to specify a model name. As I don't like calling my features `gender-model1`, `gender-model2`, ...

Reply:
>  The question is then how to get <model>. One way is that a user might be providing it by an attribute or an argument when instantiating the audinterafce object. 

Then we again have the problem that the user accidentally provides the wrong name and will not notice that the wrong cache file is loaded.

Reply:
> I would also skip the additional test, but I would like to have the possibility to specify a model name. As I don't like calling my features gender-model1, gender-model2, ...

Sorry, but I think I don't get your point.

Reply:
Yes, we cannot avoid this.

But that is the reason why I proposed to set the name automatically, when using `audmodel` or `audonnx`.

Another advantage of not storing everything in the same folder is that it will become slightly easier to clean up the cache after a while.

Reply:
But you don't have to store it in the same folder. The way I see it is that with `model1` you store it in `/cache/model1` and with `model2` in `/cache/model2` etc.

Reply:
I just don't think we need to introduce another argument for that.

Reply:
And as I said the risk of loading the wrong cache file still exists, so it doesn't helper here either.

Reply:
> But you don't have to store it in the same folder. The way I see it is that with model1 you store it in /cache/model1 and with model2 in /cache/model2 etc.

> I just don't think we need to introduce another argument for that.

I agree.

I also misunderstood the `cache_root` argument. So at the moment we do not support using a shared cache folder automatically (in that case it would be handy to automatically get model name from `process_func.uid`), but the user has to provide the `cache_root` argument explicitly, otherwise no cache is used at all.

Maybe this is indeed the best way to start and see if it works out.
But I would then use only the index and not the feature name for hash calculation, then you can also use the cache with `audinterface.Process`.

Reply:
> But I would then use only the index and not the feature name for hash calculation, then you can also use the cache with audinterface.Process.

But that's what we already do, we only use the index to create the hash.

Reply:
So would you suggest to remove both security checks then?

Reply:
I removed the sanity checks.

Reply:
I think it makes indeed sense to remove the sanity check if the output of the model matches. The oterh one I will discuss below.